File: software/frontend/src/types/reportes.ts
// src/types/reportes.ts
export interface VReporteLecturasSensor {
  lectura_id: number
  sensor_codigo: string
  sensor_nombre: string
  lectura_valor: string
  fecha_lectura: string
}

export interface VReporteAlertasInvernadero {
  alerta_id: number
  invernadero_id: number
  invernadero_nombre: string | null
  tipo_alerta: string
  mensaje: string
  fecha_generacion: string
}

export interface VReporteInventarioDispositivos {
  dispositivo_id: number
  codigo: string
  nombre: string
  tipo_dispositivo: string
  estado_dispositivo_id: number
  estado_dispositivo_nombre: string | null
}

export interface VReporteRiegoEjecutado {
  decision_id: number
  invernadero_id: number
  invernadero_nombre: string | null
  texto_decision: string
  inicio_evento: string
  duracion_segundos: number | null
}

export interface VReportePrediccionesAgua {
  prediccion_id: number
  fuente_agua: string
  modelo_usado: string
  fecha_objetivo: string
  volumen_predicho_l: string
}

export type ReportType =
  | 'lecturas'
  | 'alertas'
  | 'inventario'
  | 'riego'
  | 'predicciones'

export interface BaseFilters {
  skip: number
  limit: number
}

export interface LecturasFilters extends BaseFilters {
  q: string
  sensor_codigo: string
  sensor_nombre: string
  fecha_desde: string
  fecha_hasta: string
}

export interface AlertasFilters extends BaseFilters {
  q: string
  tipo_alerta: string
  mensaje: string
  invernadero_id: string
  fecha_desde: string
  fecha_hasta: string
}

export interface InventarioFilters extends BaseFilters {
  q: string
  codigo: string
  nombre: string
  tipo_dispositivo: string
  estado_dispositivo_id: string
}

export interface RiegoFilters extends BaseFilters {
  q: string
  invernadero_id: string
  texto_decision: string
  fecha_desde: string
  fecha_hasta: string
}

export interface PrediccionesFilters extends BaseFilters {
  q: string
  fuente_agua: string
  modelo_usado: string
  fecha_desde: string
  fecha_hasta: string
}

export const defaultLecturasFilters: LecturasFilters = {
  q: '',
  sensor_codigo: '',
  sensor_nombre: '',
  fecha_desde: '',
  fecha_hasta: '',
  skip: 0,
  limit: 100,
}

export const defaultAlertasFilters: AlertasFilters = {
  q: '',
  tipo_alerta: '',
  mensaje: '',
  invernadero_id: '',
  fecha_desde: '',
  fecha_hasta: '',
  skip: 0,
  limit: 100,
}

export const defaultInventarioFilters: InventarioFilters = {
  q: '',
  codigo: '',
  nombre: '',
  tipo_dispositivo: '',
  estado_dispositivo_id: '',
  skip: 0,
  limit: 100,
}

export const defaultRiegoFilters: RiegoFilters = {
  q: '',
  invernadero_id: '',
  texto_decision: '',
  fecha_desde: '',
  fecha_hasta: '',
  skip: 0,
  limit: 100,
}

export const defaultPrediccionesFilters: PrediccionesFilters = {
  q: '',
  fuente_agua: '',
  modelo_usado: '',
  fecha_desde: '',
  fecha_hasta: '',
  skip: 0,
  limit: 100,
}

File: software/API/app/schemas/reportes.py
# app/schemas/reportes.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


# =========================================================
# BASE
# =========================================================

class ORMBaseSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)


# =========================================================
# SINCRONIZACIONES MCP
# =========================================================

class SincronizacionMCPBase(BaseModel):
    estado_sincronizacion_id: int = Field(..., ge=1)
    dispositivo_id: int | None = Field(default=None, ge=1)
    origen: str = Field(..., min_length=1, max_length=50)
    destino: str = Field(..., min_length=1, max_length=50)
    tipo_recurso: str = Field(..., min_length=1, max_length=50)
    cantidad_registros: int = Field(default=0, ge=0)
    fecha_fin: datetime | None = None
    mensaje_resultado: str | None = Field(default=None, max_length=500)


class SincronizacionMCPCreate(SincronizacionMCPBase):
    pass


class SincronizacionMCPUpdate(BaseModel):
    estado_sincronizacion_id: int | None = Field(default=None, ge=1)
    dispositivo_id: int | None = Field(default=None, ge=1)
    origen: str | None = Field(default=None, min_length=1, max_length=50)
    destino: str | None = Field(default=None, min_length=1, max_length=50)
    tipo_recurso: str | None = Field(default=None, min_length=1, max_length=50)
    cantidad_registros: int | None = Field(default=None, ge=0)
    fecha_fin: datetime | None = None
    mensaje_resultado: str | None = Field(default=None, max_length=500)


class SincronizacionMCPResponse(SincronizacionMCPBase, ORMBaseSchema):
    id: int
    fecha_inicio: datetime
    estado_sincronizacion_nombre: str | None = None
    dispositivo_codigo: str | None = None
    dispositivo_nombre: str | None = None


# =========================================================
# VISTAS PARA REPORTES
# =========================================================

class VReporteLecturasSensorResponse(ORMBaseSchema):
    lectura_id: int
    sensor_codigo: str
    sensor_nombre: str
    lectura_valor: Decimal
    fecha_lectura: datetime


class VReporteAlertasInvernaderoResponse(ORMBaseSchema):
    alerta_id: int
    invernadero_id: int
    invernadero_nombre: str | None = None
    tipo_alerta: str
    mensaje: str
    fecha_generacion: datetime


class VReporteInventarioDispositivosResponse(ORMBaseSchema):
    dispositivo_id: int
    codigo: str
    nombre: str
    tipo_dispositivo: str
    estado_dispositivo_id: int
    estado_dispositivo_nombre: str | None = None


class VReporteRiegoEjecutadoResponse(ORMBaseSchema):
    decision_id: int
    invernadero_id: int
    invernadero_nombre: str | None = None
    texto_decision: str
    inicio_evento: datetime
    duracion_segundos: int | None = Field(default=None, ge=0)


class VReportePrediccionesAguaResponse(ORMBaseSchema):
    prediccion_id: int
    fuente_agua: str
    modelo_usado: str
    fecha_objetivo: date
    volumen_predicho_l: Decimal


# =========================================================
# ESTADOS DE SINCRONIZACIÓN
# =========================================================

class EstadoSincronizacionBase(BaseModel):
    nombre: str = Field(..., min_length=1, max_length=30)
    descripcion: str | None = Field(default=None, max_length=255)


class EstadoSincronizacionCreate(EstadoSincronizacionBase):
    pass


class EstadoSincronizacionUpdate(BaseModel):
    nombre: str | None = Field(default=None, min_length=1, max_length=30)
    descripcion: str | None = Field(default=None, max_length=255)


class EstadoSincronizacionResponse(EstadoSincronizacionBase, ORMBaseSchema):
    id: int


# =========================================================
# REPORTES SEMANALES
# =========================================================

class ReporteSemanalBase(BaseModel):
    periodo_inicio: date
    periodo_fin: date
    volumen_captado_l: Decimal = Field(default=Decimal("0.0"))
    volumen_predicho_l: Decimal = Field(default=Decimal("0.0"))
    eficiencia_riego: Decimal | None = None
    total_alertas: int = Field(default=0, ge=0)
    resumen: str | None = Field(default=None, max_length=2000)
    generado_por_ci: str | None = Field(default=None, min_length=1, max_length=20)


class ReporteSemanalCreate(ReporteSemanalBase):
    pass


class ReporteSemanalUpdate(BaseModel):
    periodo_inicio: date | None = None
    periodo_fin: date | None = None
    volumen_captado_l: Decimal | None = None
    volumen_predicho_l: Decimal | None = None
    eficiencia_riego: Decimal | None = None
    total_alertas: int | None = Field(default=None, ge=0)
    resumen: str | None = Field(default=None, max_length=2000)
    generado_por_ci: str | None = Field(default=None, min_length=1, max_length=20)


class ReporteSemanalResponse(ReporteSemanalBase, ORMBaseSchema):
    id: int
    fecha_generacion: datetime